### ISSUE TITLE
[Bug] scatter_add_ 2D fast path ignores non-contiguous strides and triggers HSA VMFault

### ISSUE THREAD

**Issue by luojung** [FlagOpen/FlagGems#6017]
## Describe the bug

FlagGems' 2D `scatter_add_` fast path does not honor the actual strides of `index`, `src`, and the output tensor. A zero-stride broadcasted `index` can therefore be read out of bounds and may cause an HSA VMFault on Hygon DCU. Other non-contiguous layouts can silently produce incorrect results.

This was reproduced in eager mode on September 4, 2026. No `torch.compile`, Dynamo, AOTAutograd, or Inductor graph was involved.

### Minimal reproduction

The reproducer uses only `torch_fl`, `torch`, and `flag_gems`; `torch_scatter` and DiffDock are not required.

```python
import torch_fl  # must be imported before torch in this environment
import torch
import flag_gems

rows, cols, nodes = 1280, 24, 128
device = flag_gems.device

src = torch.randn((rows, cols), device=device, dtype=torch.float32)
base_index = torch.arange(rows, device=device, dtype=torch.long) % nodes
index = base_index.view(rows, 1).expand(rows, cols)  # stride == (1, 0)
out = torch.zeros((nodes, cols), device=device, dtype=torch.float32)

expected = torch.zeros((nodes, cols), dtype=torch.float32).scatter_add_(
    0, index.cpu(), src.cpu()
)
print("src:", src.shape, src.stride())
print("index:", index.shape, index.stride())

with flag_gems.use_gems():
    actual = out.scatter_add_(0, index, src)

print("max diff:", (actual.cpu() - expected).abs().max().item())
```

### Expected behavior

The call should complete and match native PyTorch. A broadcasted view with stride `(1, 0)` is a valid input layout.

### Actual behavior

On Hygon DCU, the reproducer can terminate the process with exit code 134:

```text
HSA_STATUS_ERROR_MEMORY_APERTURE_VIOLATION
KERNEL VMFault
kernel name: scatter_add_2d_kernel
exit code: 134
```

A layout matrix on the unmodified implementation produced the following results:

| Layout | `src` stride | `index` stride | output stride | Result |
|---|---:|---:|---:|---|
| all contiguous | `(24, 1)` | `(24, 1)` | `(24, 1)` | pass, max diff 0 |
| broadcast index | `(24, 1)` | `(1, 0)` | `(24, 1)` | fail, max diff 19270 / may VMFault |
| row-sliced src | `(48, 1)` | `(24, 1)` | `(24, 1)` | pass, max diff 0 |
| column-sliced src | `(48, 2)` | `(24, 1)` | `(24, 1)` | fail, about 95.8% elements incorrect |
| non-contiguous output | `(24, 1)` | `(24, 1)` | `(48, 1)` | fail, about 99.2% elements incorrect |

Making only `src` contiguous does not fix the broadcast-index case. Making the expanded `index` contiguous avoids this specific trigger, but allocates a full `(rows, cols)` int64 copy and does not fix non-unit `src.stride(1)` or a non-contiguous output.

## Environment

```text
Accelerator: Hygon DCU
Device type: flagos (PrivateUse1)
Python: 3.10
PyTorch metadata/runtime: 2.10.0+cpu (Torch-FL official frontend / DCU boxing environment)
DTK: 26.04
FLAGOS_USE_FLAGGEMS: enabled for the failing path
Execution mode: eager
```

The reproduced `src/flag_gems/ops/scatter_add.py` has SHA256:

```text
88a67a2640ca0f522045df139431e18016e2d7dc36d77fb6eff488d61acbd163
```

This is identical to the file on current `master` commit `f678d805601f4622639b10376ae98da4aceefba3` inspected on September 4, 2026.

## Root-cause evidence

For `inp.ndim == 2` and `index.numel() <= 131072`, `scatter_add_0` selects `scatter_add_2d_kernel`.

The kernel currently calculates:

```python
row = offsets // idx_ncols
col = offsets % idx_ncols
idx_offsets = row * idx_ncols + col
src_offsets = row * src_stride0 + col

if DIM == 0:
    out_offsets = idx * out_ncols + col
else:
    out_offsets = row * out_ncols + idx
```

These formulas implicitly require:

```text
index.stride() == (idx_ncols, 1)
src.stride(1) == 1
out.stride() == (out_ncols, 1)
```

However, the fast-path guard does not check these conditions and does not make the tensors contiguous.

In the minimal reproducer:

```text
index shape   = (1280, 24)
index stride  = (1, 0)
backing storage contains only 1280 int64 elements
```

The correct index address is:

```python
row * 1 + col * 0
```

The current kernel instead uses:

```python
row * 24 + col
```

It may therefore read up to logical offset 30719 from storage containing only 1280 elements. The invalid value is then used to calculate an output address, which explains both incorrect results and the device VMFault.

The verified execution chain is:

```text
Tensor.scatter_add_(dim, expanded_index, src)
  -> Torch-FL / FlagGems operator routing
  -> flag_gems.ops.scatter_add.scatter_add_0
  -> scatter_add_2d_kernel
  -> DCU atomic_add
```

Disabling FlagGems before the first `import torch_fl` makes the same workload pass through Torch-FL pure boxing:

```bash
unset FLAGOS_USE_FLAGGEMS
python repro.py
```

## Suggested fix

Please keep the contiguous path fast while adding a stride-aware path.

Recommended design:

1. At the Python launch site, detect the layout:

   ```python
   simple_layout = (
       index.is_contiguous()
       and src_strided.stride(1) == 1
       and out.is_contiguous()
   )
   ```

2. Pass `SIMPLE_LAYOUT` as a `tl.constexpr` kernel argument.
3. In the `SIMPLE_LAYOUT=True` branch, retain the existing address formulas.
4. In the `SIMPLE_LAYOUT=False` branch, pass and use the complete 2D strides:

   ```python
   idx_offsets = row * index_stride0 + col * index_stride1
   src_offsets = row * src_stride0 + col * src_stride1

   if DIM == 0:
       out_offsets = idx * out_stride0 + col * out_stride1
   else:
       out_offsets = row * out_stride0 + idx * out_stride1
   ```

5. Keep runtime stride values in `do_not_specialize` so the cache does not specialize for every stride value. This should create only two layout specializations.

An alternative is to preserve the existing kernel for the contiguous path and add a separate stride-aware kernel selected at the Python launch site.

A prototype that unconditionally used the full runtime strides fixed correctness on Hygon DCU but caused a repeatable regression on contiguous inputs, so I am reporting the issue rather than proposing that prototype as a PR:

| Contiguous case | Current median | All-stride prototype median | Change |
|---|---:|---:|---:|
| dim0, `257 x 24` | 106.75 us | 110.98 us | +3.96% |
| dim0, `512 x 256` | 109.02 us | 112.45 us | +3.15% |
| dim1, `257 x 24` | 107.92 us | 109.60 us | +1.56% |
| dim1, `512 x 256` | 70.68 us | 71.38 us | +0.99% |

The likely reason is that the unconditional version adds runtime stride arguments and integer multiplications to every contiguous invocation, preventing constant folding of the simple layout.

## Suggested regression coverage

Please cover at least:

- `dim=0` and `dim=1`;
- `float32`, `float16`, and `bfloat16` where supported;
- zero-stride broadcasted index;
- row/column sliced and transposed index;
- row/column sliced and transposed src;
- row/column sliced and transposed output;
- the `N == 131072` fast-path boundary and `N > 131072` generic path;
- in-place return identity, data pointer, stride, and version counter;
- storage holes in non-contiguous outputs remaining unchanged;
- contiguous-path performance on all common backends using this kernel.

A private correctness prototype passed 70 layout/dimension/dtype/boundary cases and direct `torch_scatter` forward/backward comparisons on Hygon DCU, with max difference 0. It has not been proposed as a PR because the contiguous performance regression above still needs to be eliminated and other vendor backends require CI validation.

## Duplicate search

I searched open and closed FlagGems issues and pull requests for `scatter_add_`, `scatter_add_2d_kernel`, `VMFault`, `HSA_STATUS_ERROR_MEMORY_APERTURE_VIOLATION`, non-contiguous scatter, and broadcast-index stride reports. No matching report was found as of September 4, 2026.
